### ISSUE TITLE
IP ban investigation

### ISSUE THREAD

**Issue by sigaloid** [redlib-org/redlib#318]
## Update (11/19): #324 is **solved**. Please do not comment here about a JSON error unless you are **certain** that you're on the latest version; the widespread outage was unrelated to the issue of IP bans.

There is a number of reports of continued JSON errors despite the fact that most instance operators running on clean IPs do not get these. If you are impacted by this, please ensure you can reproduce this on certain IPs by confirming you can visit the Redlib home page on one IP (perhaps try your home IP, VPN to another place, etc) but cannot on the IP in question. Then, comment or email (ipban @ my domain, linked in profile) the following info:

* Confirm you're running the latest commit (you should be able to see "✅ Instance is up to date" on the error page)
* Confirm impact (can you view any pages?)
* IPv4/IPv6
* Your IP's ASN and ISP (and IP if you're comfortable sharing privately)

### Reminder: Do not comment on this post if you're getting errors, unless you've **CONFIRMED** it works on some IP that isn't yours.

**Comment by ggtylerr** [comment]
* Confirmed.
* Occasional blockage, ~75-90% of runtime is unblocked.
* NYC-1: `45.137.206.17`, CAL-1: `71.19.146.127`
  - (IPv6 ranges can be disclosed in PGP email.)
* NYC-1: RoyaleHosting BV, ASN `212477`. CAL-1: PRGMR.com Inc, ASN `47066`.

It should be noted that both are currently fully operational but has previously been experiencing #301.

**Comment by sigaloid** [comment]
Can you set the environment variable `RUST_LOG=redlib=trace`? Then when you encounter the error, excerpt from the latest logs? Thanks.

**Comment by hyperreal64** [comment]
- Confirmed.
- I cannot view any pages.
- IPv4: 152.53.37.179
- IPv6: 2a0a:4cc0:2000:2a:1416:76ff:fe0c:d737
- netcup GmbH ASN: 214996

**Comment by ggtylerr** [comment]
> Can you set the environment variable `RUST_LOG=redlib=trace`? Then when you encounter the error, excerpt from the latest logs? Thanks.

Just got the error on both servers. It looks like NYC-1 just became 3 commits too old due to unrelated issues regarding quay.io (it looks like their IPv6 connection is down and `docker compose pull` is forcing connections over that.) However CAL-1 is on latest. The logs, unfortunately, don't seem to be of much use:

```
redlib  |  WARN  redlib::client > Rate limit 9 is low. Spawning force_refresh_token()           redlib  |  TRACE redlib::oauth  > Rolling over refresh token. Current rate limit: 8             redlib  |  INFO  redlib::oauth  > [🔄] Spoofing Android client with headers: {"Client-Vendor-Id": "d5bc0918-e77d-40f6-9ef0-69274c59539a", "X-Reddit-Device-Id": "d5bc0918-e77d-40f6-9ef0-69274c59539a", "User-Agent": "Reddit/Version 2024.04.0/Build 1391236/Android 9"}, uuid: "d5bc0918-e77d-40f6-9ef0-69274c59539a", and OAuth ID "ohXpoqrZYub1kg"                                          redlib  |  TRACE redlib::oauth  > Sending token request...                                      redlib  |  ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 403 Forbidden                                                             redlib  |  ERROR redlib::utils  > Error page rendered: Failed to parse page JSON data: expected value at line 1 column 1                                                                        redlib  |  ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 403 Forbidden                                                             redlib  |  ERROR redlib::utils  > Error page rendered: Failed to parse page JSON data: expected value at line 1 column 1                                                                        redlib  |  TRACE redlib::oauth  > Received response with status 200 OK and length Some("1308")  redlib  |  TRACE redlib::oauth  > Serializing response...                                       redlib  |  TRACE redlib::oauth  > Accessing relevant fields...                                  redlib  |  INFO  redlib::oauth  > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399                                                                        redlib  |  INFO  redlib::oauth  > [✅] Successfully created OAuth client                        redlib  |  ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 403 Forbidden                                                             redlib  |  ERROR redlib::utils  > Error page rendered: Failed to parse page JSON data: expected value at line 1 column 1                                                                        redlib  |  ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 403 Forbidden
```

**Comment by sigaloid** [comment]
Hm, the fact that it's near to a token refresh makes me wonder if it's the fact that the token reaches its rate limit too quickly. It currently starts a background task when there's 9 left, which is working: 

![image](https://github.com/user-attachments/assets/87983ba8-4397-453b-9258-fdf6f674dd19)


But... if 9 concurrent requests come in simultaneously, it's possible they exhaust the remaining rate limit count. Is this a high-traffic instance, specifically?

**Comment by ggtylerr** [comment]
> Is this a high-traffic instance, specifically?

Definitely not. CAL-1 is one of two secondary servers, only listed on my site. It's primarily used for Invidious.

**Comment by sigaloid** [comment]
Hm, yeah, that checks out. So far every IP has been a commercial ASN and I'm hoping that's not the commonality, that they all have flagged some heuristic for being suspicious because they're hosted.

**Comment by ggtylerr** [comment]
I am noting an unusual amount of requests on NYC-1 (the public server on redlib's list.) I've heavily adjusted rate limits over nginx to be only 5 requests per second (10 burst) but without any further logging, it's difficult to say whether this is regular user activity or possibly scrapers.
 https://github.com/user-attachments/assets/8ed8eb93-d0f5-430d-89bb-93bc830ab8e2

**Comment by np22-jpg** [comment]
I was running the same commit and couldn’t view any pages, either, on a residential IP (which is why I’m not comfortable with sharing it). I was able to get around it by taking my instance offline, requesting an IP unban from Reddit, and bringing it back up. 

I'm not sure if this is at all helpful information.

**Comment by sigaloid** [comment]
That's very helpful. How did you request an unban? And did you have the same IP the whole time (during the ban and after it worked again)? Sometimes residential IPs cycle and change normally.

**Comment by np22-jpg** [comment]
When I got IP banned, I received a page that looked like [this](https://www.blackhatworld.com/attachments/reddiot-png.355987/) or [this](https://preview.redd.it/firefox-your-request-has-been-blocked-due-to-a-network-v0-d3o41256ynud1.png?width=1080&crop=smart&auto=webp&s=ee657d4735986cd28777a0c07517026ee68c00e9) when trying to access from my browser. Based on my previous emails, the link to contact support led me to ratelimit@reddit.zendesk.com, who then pointed me to fill out [this form](https://support.reddithelp.com/hc/en-us/requests/new?ticket_form_id=21879292693140). I have had the same IP the entire time.

**Comment by sigaloid** [comment]
Weird. After only a few hours, it should have only requested one token total and presumably not had enough traffic (more than 99 reqs in a 5 minute period) to require a new token rollover. That's really surprising that's all it took to get your IP under their watchful eye. I'm going to think a bit on how to proceed here.  I don't know if it makes sense to attempt to identify a single factor that causes this policy ban, at least via trial and error, for a few reasons:

* it could be any number of factors
* it could be ML-type ban of many factors in which case we really couldn't guess and check over and over
* I don't have unlimited clean IPs that I can test with
* I already put a lot of work into attempting to copy exact behavior of the mobile apps in terms of new tokens, it's unclear if it's a single thing I missed or if it's some lower level thing like TLS signatures, etc...

Maybe I need to take a look at the auth flow on a current app since it's been a few months. Not sure if anything that has changed should really cause this, since people who haven't used the app in a few months should still be able to use the app without being IP banned.

**Comment by hyperreal64** [comment]
OK, right now I can access Redlib from my home IP instance. My public-facing instance is still error.

**Comment by Shagon94** [comment]
Seeing the same issue again for some reason, I'm self-hosting it and the only one interacting with the instance (single user)

**Comment by donslice** [comment]
Getting this as well on my private (single user) self-hosted instance. 

Issue started at 2:25 PM Eastern Time. I can access the website just fine

`redlib  | Starting Redlib...
redlib  |  INFO  redlib > Evaluating config.
redlib  |  INFO  redlib > Evaluating instance info.
redlib  |  INFO  redlib > Creating OAUTH client.
redlib  |  INFO  redlib::oauth > [🔄] Spoofing Android client with headers: {"User-Agent": "Reddit/Version 2023.21.0/Build 956283/Android 13", "X-Reddit-Device-Id": "0511a00b-c160-4a0c-8b26-1e47ac7b12e9", "Client-Vendor-Id": "0511a00b-c160-4a0c-8b26-1e47ac7b12e9"}, uuid: "0511a00b-c160-4a0c-8b26-1e47ac7b12e9", and OAuth ID "ohXpoqrZYub1kg"
redlib  |  TRACE redlib::oauth > Sending token request...
redlib  |  TRACE redlib::oauth > Received response with status 200 OK and length Some("1308")
redlib  |  TRACE redlib::oauth > Serializing response...
redlib  |  TRACE redlib::oauth > Accessing relevant fields...
redlib  |  INFO  redlib::oauth > [✅] Success - Retrieved token "eyJhbGciOiJSUzI1NiIsImtpZCI6IlNI...", expires in 86399
redlib  |  INFO  redlib::oauth > [✅] Successfully created OAuth client
redlib  |  INFO  redlib::oauth > [⏳] Waiting for 86279s seconds before refreshing OAuth token...
redlib  | Running Redlib v0.35.1 on [::]:8080!
redlib  |  ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 403 Forbidden
redlib  |  ERROR redlib::utils  > Error page rendered: Failed to parse page JSON data: expected value at line 1 column 1 
redlib  |  ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 403 Forbidden
redlib  |  ERROR redlib::utils  > Error page rendered: Failed to parse page JSON data: expected value at line 1 column 1 `

**Comment by kotx** [comment]
- Confirmed
- No pages can be viewed (although curling reddit frontpage with both v4 and v6 somehow works?)
- `69.31.3.91`, `2605:4c40:118:ed69:0:6990:3f1c:1`
- AS30081 CacheNetworks, Inc. (hosted on Fly.io)

```
 2024-11-18T19:46:47.815 app[8749e0b0249d28] ord [info] ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 403 Forbidden

2024-11-18T19:46:47.815 app[8749e0b0249d28] ord [info] ERROR redlib::utils > Error page rendered: Failed to parse page JSON data: expected value at line 1 column 1 
```

**Comment by cameronj86** [comment]
I've been self-hosting a private teddit instance for over a year that was blocked this morning, so the issue appears to be widespread FYI.

Gave a shot @ redlib just now:

1. Confirmed
2. Only settings page can be viewed
3. ?
4. Comcast

**Comment by Cyrix126** [comment]
Got the error since a few hours.
`Failed to parse page JSON data: expected value at line 1 column 1 | /r/popular/hot.json?&raw_json=1&geo_filter=GLOBAL`

Getting on reddit without redlib works fine from same ip.
Restarting doesn't solve the issue.
This is on a self hosted instance.

TRACE log:

```
 ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 403 Forbidden
 ERROR redlib::utils  > Error page rendered: Failed to parse page JSON data: expected value at line 1 column 1
```

**Comment by grngxd** [comment]
same here, cant view any pages but only on SOME public instances

**Comment by sigaloid** [comment]
Yeah this is a widespread thing now. This issue was intended to be for actual IP bans (where it would work any other IP except one) but now there's just actually a server side change. Triaging now (3x in the last 2 months. sigh). Everyone can ignore the questions about IPs if you're here to report the current outage :smile:

**Comment by argium** [comment]
Is the process for debugging this error documented? How do I dump the underlying request/response?

**Comment by OxyMagnesium** [comment]
Given that there seem to be IP-unrelated issues also going on right now, it might also be worth adding a step to try `curl`ing `https://www.reddit.com` on the instance host to determine whether it's an IP ban. I get [this](https://www.blackhatworld.com/attachments/reddiot-png.355987/) (image courtesy of @np22-jpg) when I do that on my host as of an hour or two ago, which does indicate an IP ban.

For the other questions:
1. Confirmed.
2. All requests, whether through redlib or direct to `reddit.com` return 403.
3. IPv4.
4. It's a commercial AS. I can provide details if it'll help.

This was a public instance, with somewhere around 200k requests/day.

**Comment by maxysoft** [comment]
Yesterday my instance was working, today not anymore.. Trying to curl reddit.com:
`You've been blocked by network security.`
(but from what I remember it's always been like this)

The instance seems to be completely blocked.

- I'm on the latest version
- Tried ipv4 and ipv6
- Hetzner ASN

**Comment by tcsenpai** [comment]
I just want to add that I incur in the same problem. Is there a way to specify a proxy server (or even better something like gluetun) for redlib? I am using the standalone built binary

**Comment by sigaloid** [comment]
It's not an IP ban issue, it's a widespread problem affecting every Redlib user. Proxy won't help. #324

**Comment by maxysoft** [comment]
> It's not an IP ban issue, it's a widespread problem affecting every Redlib user. Proxy won't help. #324

I can confirm that. Just tried redlib from home (so residential ip) and still doesn't work. Seems that they're blocking redlib specifically :(

**Comment by graysonlee123** [comment]
Chiming in to say I'm getting this issue on a residential instance as well.

**Comment by sigaloid** [comment]
for all commenting about the JSON error; that's a different issue #324 which is now fixed. Update here if you're experiencing the same error AFTER pulling the LATEST docker image.

**Comment by tcsenpai** [comment]
I can confirm updating to the recently pushed version works for me too.

Specifically, from `v0.35.1` the JSON error disappears.

My step by step solution (I use redlib as a binary for a systemd service):

```
git pull
sudo systemctl stop redlib
cargo build --release
sudo rm -rf /usr/bin/redlib
sudo cp target/release/redlib /usr/bin/redlib
```

**Comment by OxyMagnesium** [comment]
Okay it seems I was wrong about direct requests to `reddit.com` being indicative of an IP ban, at least for the endpoints Redlib accesses. When Redlib is not being blocked (like now), requests through Redlib on my instance host work, but direct requests to `reddit.com` from the same host are blocked. I don't think it's a problem with `curl` either because if I proxy requests generated from my desktop's browser through my instance host, they still receive the "blocked by network security" 403 page.

Also just to echo everyone else: thank you so much @sigaloid for all your hard work 😄

**Comment by sigaloid** [comment]
> requests through Redlib on my instance host work, but direct requests to reddit.com from the same host are blocked

wait, that's huge! You mean that Redlib works even though in the browser doesn't? I think that means that Redlib no longer triggers the IP bans directly (or are being blocked by them), it's just a residual ban that will eventually expire.

**Comment by OxyMagnesium** [comment]
That's what I'm seeing, yes. Specifically:

- Residential connection:
  - `curl https://www.reddit.com` works
  - Desktop browser request to `www.reddit.com`  works
  - Redlib as of the current time and latest commit works

- Commercial (former public instance host) connection:
  - `curl https://www.reddit.com` does NOT work
  - Desktop browser request to `www.reddit.com` does NOT work
  - Redlib as of the current time and latest commit DOES work

But to be clear, it seems to me that Redlib on this host was never affected by the IP ban in the first place -- I just didn't realize that it could be blocked when requesting `www.reddit.com` while still being allowed through on the endpoints Redlib is accessing.

@maxysoft mentioned that their instance host has never been able to curl `www.reddit.com`, but Redlib worked previously. I think I might have been in the same situation without realizing, and the breakage today was purely due to the TLS fingerprinting issue. I'm not sure how the blocking of direct requests relates (or doesn't relate) to what people were seeing earlier in the thread.

**Comment by 2bc4** [comment]
> requests through Redlib on my instance host work, but direct requests to reddit.com from the same host are blocked

Confirming this here as well. If I route redlib with the latest commit through a VPN with a blocked IP it still works.

**Comment by WreckingBANG** [comment]
Everything is working for me too on my private residential instance on the latest image. 👍

**Comment by sigaloid** [comment]
Okay, it's come to my attention that there were some deeper server-side changes that are causing these IP bans. This has caused some serious issues for the popular instances as there's no rigorous rate limit bypass mechanism. The main issue is this rate limiting effect so popular instances are hit hard by them.

Unfortunately it looks like a lot of work is necessary to get around this restriction. It may be radio silence (and current status quo - local instances will probably work, but instances that reach rate limit (100 every 10 minutes) may get constant errors) until I can get a real overhauled solution up and running. I can't put a timeline on it, unfortunately. It will be possible that public instances will work, while local instances built from the repo's head won't bypass the rate limiting. Once I'm more sure of its stability, this new version will of course be published here, but only once it's in a place where I'm actually sure it's resilient to the issues it is facing now.

**Comment by donslice** [comment]
Sheesh. They really don't want us using this. As always, thanks for your continued efforts.

**Comment by sigaloid** [comment]
Issue turned out to be a lot easier to fix than my initial thought... :fearful: oops. Wanted to ensure that's all that went wrong before pushing it. We should be at a point where IP bans aren't correlated to IP's. If it still is, it's probably a unique case with your IP and it shouldn't be as widespread. If I get tons of reports about it, I can open a new issue. 

Public instance operators, please update!

**Comment by ggtylerr** [comment]
Unfortunately it looks like Reddit has blocked my NYC-1 instance again - unsure if it's an IP block or what, but CAL-1 also *occasionally* gets it.

**Comment by jaydenengrayson** [comment]
For those looking to swap faces or make adjustments to their photos, using a [free online face swap](https://airbrush.com/ai-face-swap) tool can be a great option. These tools allow for quick edits without needing complex software. Feel free to explore some free online platforms that provide these services for easy and fast modifications!

**Comment by Ruyeex** [comment]
Have the same issue.
With using Mullvad

**Comment by scriptsandthings** [comment]
Is there a max number of subreddits one can use in redlib?

Along with the failed to parse message, I also receive a "URI Too Long" error once adding additional subreddits. If I remove the subreddits from my docker container and restart it, the error goes away and redlib works again.

Not sure how many subreddits tipped it over the edge, but I was attempting to use a total of 936 subreddits when I encountered the issue. Cutting the number in half to 468 resolved the URI length message.

Assuming this limit would be on reddit's side, not anything redlib could work around without sending multiple requests after x subreddits have been added.

> redlib  |  ERROR redlib::client > Got an invalid response from reddit expected value at line 1 column 1. Status code: 414 URI Too Long
> redlib  |  ERROR redlib::utils  > Error page rendered: Failed to parse page JSON data: expected value at line 1 column 1
